src/tools/immutable_object_cache/ObjectCacheStore.cc

```diff
@@ -193,11 +243,28 @@ int ObjectCacheStore::lookup_object(std::string pool_nspace,
 
   switch (ret) {
     case OBJ_CACHE_NONE: {
-      pret = do_promote(pool_nspace, pool_id, snap_id, object_name);
-      if (pret < 0) {
-        lderr(m_cct) << "fail to start promote" << dendl;
+      if (take_token_from_throttle(object_size, 1)) {
+        pret = do_promote(pool_nspace, pool_id, snap_id, object_name);
+        if (pret < 0) {
+          lderr(m_cct) << "fail to start promote" << dendl;
+        }
+        return ret;
+      } else {
+        m_policy->update_status(cache_file_name, OBJ_CACHE_THROTTLE);
+        return OBJ_CACHE_THROTTLE;
+      }
+    }
+    case OBJ_CACHE_THROTTLE: {
+      if (take_token_from_throttle(object_size, 1)) {
+        m_policy->update_status(cache_file_name, OBJ_CACHE_SKIP);
+        pret = do_promote(pool_nspace, pool_id, snap_id, object_name);
```

Comment:
Nit: I'm not sure we need a new state here. Just because the policy said to promote an object at time T1 (but it was then throttled) doesn't mean that the object should automatically be promoted at time T2 (if the throttle has space). Instead, I would expect the policy to return the object's state to pre time T1 (before the original `do_promote` call) so that it can be evaluated again if and only if the data is re-requested and other items haven't already moved up to higher priority within the cache policy.

Reply:
> Nit: I'm not sure we need a new state here. Just because the policy said to promote an object at time T1 (but it was then throttled) doesn't mean that the object should automatically be promoted at time T2 (if the throttle has space). Instead, I would expect the policy to return the object's state to pre time T1 (before the original `do_promote` call) so that it can be evaluated again if and only if the data is re-requested and other items haven't already moved up to higher priority within the cache policy.

When throttling , just update status OBJ_CACHE_SKIP to OBJ_CACHE_NONE. In the next look_object(), return OBJ_CACHE_NONE directly. Here should be no need to delete entry here, _do_evict()_.